From 6309d93056a69843529c2477c13b50540ed5389a Mon Sep 17 00:00:00 2001
From: Thao Luong <thao.luong.uw@rvc.renesas.com>
Date: Wed, 27 May 2015 11:36:26 +0700
Subject: [PATCH 17/25] omxvideodec: Skip the video decoder negotiation when 
 the   caps is not changed

This patch avoids redundant negotiations in the video decoder class.

refs #69206
Signed-off-by: Thao Luong <thao.luong.uw@rvc.renesas.com>
---
 omx/gstomxvideodec.c | 34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index d136e4d..ff35fa3 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -82,6 +82,7 @@ static GstFlowReturn gst_omx_video_dec_handle_frame (GstVideoDecoder * decoder,
 static GstFlowReturn gst_omx_video_dec_finish (GstVideoDecoder * decoder);
 static gboolean gst_omx_video_dec_decide_allocation (GstVideoDecoder * bdec,
     GstQuery * query);
+static gboolean gst_omx_video_dec_negotiate2 (GstVideoDecoder * decoder);
 
 static GstFlowReturn gst_omx_video_dec_drain (GstOMXVideoDec * self,
     gboolean is_eos);
@@ -140,6 +141,8 @@ gst_omx_video_dec_class_init (GstOMXVideoDecClass * klass)
   video_decoder_class->finish = GST_DEBUG_FUNCPTR (gst_omx_video_dec_finish);
   video_decoder_class->decide_allocation =
       GST_DEBUG_FUNCPTR (gst_omx_video_dec_decide_allocation);
+  video_decoder_class->negotiate =
+      GST_DEBUG_FUNCPTR (gst_omx_video_dec_negotiate2);
 
   klass->cdata.type = GST_OMX_COMPONENT_TYPE_FILTER;
   klass->cdata.default_src_template_caps =
@@ -2918,3 +2921,34 @@ gst_omx_video_dec_get_property (GObject * object, guint prop_id,
       break;
   }
 }
+static gboolean
+gst_omx_video_dec_negotiate2 (GstVideoDecoder * decoder)
+{
+  GstOMXVideoDec *self;
+  GstVideoCodecState *state;
+  GstCaps *prevcaps;
+
+  self = GST_OMX_VIDEO_DEC (decoder);
+
+  state = gst_video_decoder_get_output_state (decoder);
+  if (state == NULL) {
+    GST_ERROR_OBJECT (self, "Failed to get output state");
+    return FALSE;
+  }
+
+  if (state->caps == NULL)
+    state->caps = gst_video_info_to_caps (&state->info);
+
+  prevcaps = gst_pad_get_current_caps (decoder->srcpad);
+  if (prevcaps && gst_caps_is_equal (prevcaps, state->caps)) {
+    GST_DEBUG_OBJECT (self,
+        "Skip the video decoder negotiation because the caps is not changed");
+    gst_video_codec_state_unref (state);
+    return TRUE;
+  }
+
+  gst_video_codec_state_unref (state);
+
+  return GST_VIDEO_DECODER_CLASS
+      (gst_omx_video_dec_parent_class)->negotiate (decoder);
+}
-- 
2.4.0

